drivers/sensor/icm42688/icm42688_decoder.c

```diff
@@ -0,0 +1,311 @@
+		case ICM42688_GYRO_FS_500:
+			*shift = 4;
+			return 0;
+		case ICM42688_GYRO_FS_1000:
+			*shift = 5;
+			return 0;
+		case ICM42688_GYRO_FS_2000:
+			*shift = 6;
+			return 0;
+		default:
+			return -EINVAL;
+		}
+	case SENSOR_CHAN_DIE_TEMP:
+		*shift = 9;
+		return 0;
+	default:
+		return -EINVAL;
+	}
+}
+
+static enum sensor_channel icm42688_get_channel_from_position(int pos)
+{
+	switch (pos) {
+	case 0:
+		return SENSOR_CHAN_DIE_TEMP;
+	case 1:
+		return SENSOR_CHAN_ACCEL_X;
+	case 2:
+		return SENSOR_CHAN_ACCEL_Y;
+	case 3:
+		return SENSOR_CHAN_ACCEL_Z;
+	case 4:
+		return SENSOR_CHAN_GYRO_X;
+	case 5:
+		return SENSOR_CHAN_GYRO_Y;
+	case 6:
+		return SENSOR_CHAN_GYRO_Z;
+	default:
+		return SENSOR_CHAN_MAX;
+	}
+}
+
+int icm42688_convert_raw_to_q31(struct icm42688_cfg *cfg, enum sensor_channel chan,
+				int32_t reading, q31_t *out)
+{
+	int32_t whole;
+	int32_t fraction;
+	int64_t intermediate;
+	int8_t shift;
+	int rc;
+
+	rc = icm42688_get_shift(chan, cfg->accel_fs, cfg->gyro_fs, &shift);
+	if (rc != 0) {
+		return rc;
+	}
+
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_XYZ:
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+		icm42688_accel_ms(cfg, reading, &whole, &fraction);
+		break;
+	case SENSOR_CHAN_GYRO_XYZ:
+	case SENSOR_CHAN_GYRO_X:
+	case SENSOR_CHAN_GYRO_Y:
+	case SENSOR_CHAN_GYRO_Z:
+		icm42688_gyro_rads(cfg, reading, &whole, &fraction);
+		break;
+	case SENSOR_CHAN_DIE_TEMP:
+		icm42688_temp_c(reading, &whole, &fraction);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+	intermediate = ((int64_t)whole * INT64_C(1000000) + fraction);
+	if (shift < 0) {
+		intermediate = intermediate * INT32_MAX * (1 << -shift) /  INT64_C(1000000);
+	} else if (shift > 0) {
+		intermediate = intermediate * INT32_MAX / (((1 << shift) - 1) * INT64_C(1000000));
+	}
+	*out = CLAMP(intermediate, INT32_MIN, INT32_MAX);
+
+	return 0;
+}
+
+static int icm42688_get_channel_position(enum sensor_channel chan)
+{
+	switch (chan) {
+	case SENSOR_CHAN_DIE_TEMP:
+		return 0;
+	case SENSOR_CHAN_ACCEL_XYZ:
+	case SENSOR_CHAN_ACCEL_X:
+		return 1;
+	case SENSOR_CHAN_ACCEL_Y:
+		return 2;
+	case SENSOR_CHAN_ACCEL_Z:
+		return 3;
+	case SENSOR_CHAN_GYRO_XYZ:
+	case SENSOR_CHAN_GYRO_X:
+		return 4;
+	case SENSOR_CHAN_GYRO_Y:
+		return 5;
+	case SENSOR_CHAN_GYRO_Z:
+		return 6;
+	default:
+		return 0;
+	}
+}
+
+static uint8_t icm42688_encode_channel(enum sensor_channel chan)
+{
+	uint8_t encode_bmask = 0;
+
+	switch (chan) {
+	case SENSOR_CHAN_DIE_TEMP:
+	case SENSOR_CHAN_ACCEL_X:
+	case SENSOR_CHAN_ACCEL_Y:
+	case SENSOR_CHAN_ACCEL_Z:
+	case SENSOR_CHAN_GYRO_X:
+	case SENSOR_CHAN_GYRO_Y:
+	case SENSOR_CHAN_GYRO_Z:
+		encode_bmask = BIT(icm42688_get_channel_position(chan));
+		break;
+	case SENSOR_CHAN_ACCEL_XYZ:
+	case SENSOR_CHAN_GYRO_XYZ:
+		encode_bmask = BIT(icm42688_get_channel_position(chan - 3)) |
+			       BIT(icm42688_get_channel_position(chan - 2)) |
+			       BIT(icm42688_get_channel_position(chan - 1));
```

Comment:
While this works, I had to dig a bit to confirm this (again) - `ACCEL_XYZ` comes right after the individual `ACCEL` sensor channel enums.  If for any reason the ordering of `enum sensor_channel` changes, this will break and may be a headache to root cause. For future safety maybe we can break this up into something like
```
	case SENSOR_CHAN_ACCEL_XYZ:
		encode_bmask = BIT(icm42688_get_channel_position(SENSOR_CHAN_ACCEL_X)) |
			       BIT(icm42688_get_channel_position(SENSOR_CHAN_ACCEL_Y)) |
			       BIT(icm42688_get_channel_position(SENSOR_CHAN_ACCEL_Z));
                break;
	case SENSOR_CHAN_GYRO_XYZ:
		encode_bmask = BIT(icm42688_get_channel_position(SENSOR_CHAN_GYRO_X)) |
			       BIT(icm42688_get_channel_position(SENSOR_CHAN_GYRO_Y)) |
			       BIT(icm42688_get_channel_position(SENSOR_CHAN_GYRO_Z));
                break;
```

Reply:
Done